cachi2/core/package_managers/yarn/resolver.py

```diff
@@ -0,0 +1,116 @@
+"""
+Resolve the dependency list for a yarn project.
+
+It also performs the necessary validations to avoid allowing an invalid
+project to keep being processed.
+"""
+import json
+import logging
+from typing import NamedTuple
+
+from cachi2.core.models.sbom import Component
+from cachi2.core.package_managers.yarn.locators import Locator, parse_locator
+from cachi2.core.package_managers.yarn.utils import run_yarn_cmd
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+
+
+class Cache(NamedTuple):
+    """Cache information for a package."""
+
+    checksum: str
+    path: str
+
+
+class Package(NamedTuple):
+    """A package listed by the yarn info command.
+
+    See the output for 'yarn info -AR --json --cache'.
+
+    Note that the attributes present in the json vary a little depending on the
+    type of dependency.
+    """
+
+    cache: Cache
+    locator: Locator
+    version: str
+
+    @classmethod
+    def from_info_string(cls, info: str) -> "Package":
+        """Create a Package from the output of yarn info."""
+        data = json.loads(info)
+        cache = data["children"]["Cache"]
+
+        return Package(
+            cache=Cache(checksum=cache["Checksum"], path=cache["Path"]),
+            locator=parse_locator(data["value"]),
+            version=data["children"]["Version"],
+        )
+
+
+def resolve_dependency_info(source_dir: RootedPath) -> list[Package]:
+    """Fetch and parse dependency info from the 'yarn info' output.
+
+    This function also performs a validation to ensure that the current
+    yarn project can be processed.
+
+    :raises PackageRejected: if the validation fails.
+    :raises SubprocessCallError: if the 'yarn info' command fails.
+    """
+    cmd = ["info", "-AR", "--json", "--cache"]
+
+    result = run_yarn_cmd(cmd, source_dir)
+    packages = [Package.from_info_string(info) for info in result.splitlines()]
+
+    _vet_git_dependencies(packages)
+
+    return packages
+
+
+def create_component_from_package(package: Package) -> Component:
+    """Create a SBOM component from a yarn Package."""
+    name = _resolve_package_name(package)
+
+    return Component(
+        name=name,
+        version=package.version,
+        purl=_generate_purl_for_package(package, name),
+    )
+
+
+def _vet_git_dependencies(packages: list[Package]) -> None:
+    """Vet a Git dependency if it contains forbidden lifecycle scripts.
+
+    :raises PackageRejected: if a dependency containing a forbidden lifecycle
+    script is found.
+    """
+    pass
+
+
+def _resolve_package_name(package: Package) -> str:
+    """Resolve the package's real name based on its protocol.
+
+    Non-registry deps might have names in their locator string that
+    differ from the one in their package.json file.
+
+    Look at the package.json name for every non-registry dependency.
+    """
+    if package.locator.protocol == "npm":
+        return package.locator.name
```

Comment:
`if isinstance(package.locator, NpmLocator):` is probably better (otherwise you have to give `protocol` and `name` to every locator type even when it doesn't make sense)

Reply:
The initial reasoning was that a Locator could have more than one protocol (GitLocator: https and ssh, for instance). I thought this would affect the purl generation, but it seems that is not the case. 

My latest changes already went for you approach (although I used `type(package.locator)`, which is a worse way of dealing with it).